tests/cloud/test_grouping.py

```diff
@@ -0,0 +1,129 @@
+
+        self.fake_tenant1 = mock.Mock()
+        self.fake_tenant1.id = 't1'
+        self.fake_tenant2 = mock.Mock()
+        self.fake_tenant2.id = 't2'
+
+        self.identity.get_tenants_list.return_value = [self.fake_tenant1,
+                                                       self.fake_tenant2]
+
+        self.fake_network_1 = {'name': 'net1'}
+        self.fake_network_2 = {'name': 'net3'}
+
+        self.network.get_networks_list.return_value = [self.fake_network_1,
+                                                       self.fake_network_2]
+
+        self.fake_instance1 = mock.Mock()
+        self.fake_instance1.id = 's1'
+        self.fake_instance2 = mock.Mock()
+        self.fake_instance2.id = 's2'
+        self.fake_instance3 = mock.Mock()
+        self.fake_instance3.id = 't3'
+
+        self.cloud = mock.Mock()
+        self.cloud().resources = {'network': self.network,
+                                  'compute': self.compute,
+                                  'identity': self.identity}
+
+        self.cloud_patch = mockpatch.PatchObject(cloud, 'Cloud',
+                                                 new=self.cloud)
+        self.useFixture(self.cloud_patch)
+
+    def tenant_side_effect(self, **kwargs):
+        if kwargs.get('tenant_id') == 't1':
+            return [self.fake_instance1, self.fake_instance3]
+        elif kwargs.get('tenant_id') == 't2':
+            return [self.fake_instance2]
+
+    def network_side_effect(self, **kwargs):
+        if kwargs.get('network_name') == 'net1':
+            return [self.fake_instance1, self.fake_instance3]
+        elif kwargs.get('network_name') == 'net3':
+            return [self.fake_instance2]
+
+    def make_group_file(self, group_rules):
+        group_file = open(FILE_NAME, 'w')
+        group_file.write(group_rules)
+
+    def test_group_by_tenant(self):
+        group_rules = """
+        group_by:
+            - tenant
+        """
+
+        self.make_group_file(group_rules)
+        group = grouping.Grouping(FAKE_CONFIG, FILE_NAME, 'src')
+        group.compute.get_instances_list.side_effect = self.tenant_side_effect
+
+        group.group()
+
+        expected_result = {'t2': ['s2'], 't1': ['s1', 't3']}
+
+        result = group.get_group_conf(RESULT_FILE)
+        self.assertEquals(expected_result, result)
+
+    def test_group_by_network(self):
+        group_rules = """
+        group_by:
+            - network
+        """
+
+        self.make_group_file(group_rules)
+        group = grouping.Grouping(FAKE_CONFIG, FILE_NAME, 'src')
+        group.compute.get_instances_list.side_effect = self.network_side_effect
+        group.group()
+
+        expected_result = {'net1': ['s1', 't3'], 'net3': ['s2']}
+
+        result = group.get_group_conf(RESULT_FILE)
+        self.assertEquals(expected_result, result)
+
+    def tearDown(self):
+        super(GroupingTestCase, self).tearDown()
+        os.remove(FILE_NAME)
+        os.remove(RESULT_FILE)
```

Comment:
Not sure if this is covered here, but please consider adding tests to verify invalid user input: non-existing VM IDs, incorrect grouping options passed, etc.